IdentityCore/src/oauth2/MSIDTokenResponse.m

```diff
@@ -30,83 +30,66 @@
+                              error:(NSString *)error
+                   errorDescription:(NSString *)errorDescription
+                          initError:(NSError **)initError
 {
-    self = [self initWithJSONDictionary:json error:error];
+    self = [super init];
     if (self)
     {
-        if (token && [NSString msidIsStringNilOrBlank:self.refreshToken])
+        _accessToken = accessToken;
+        _refreshToken = refreshToken;
+        _expiresIn = expiresIn;
+        _tokenType = tokenType;
+        _scope = scope;
+        _state = state;
+        _idToken = idToken;
+        self.additionalServerInfo = additionalServerInfo;
+        _error = error;
+        _errorDescription = errorDescription;
+        
+        NSError *localError;
+        [self initIdToken:&localError];
+        if (localError)
         {
-            self.refreshToken = token.refreshToken;
+            MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, nil, @"Failed to init id token wrapper, error: %@", MSID_PII_LOG_MASKABLE(localError));
+            if (initError) *initError = localError;
         }
     }
     
     return self;
 }
 
-- (id)initWithJSONDictionary:(NSDictionary *)json error:(NSError *__autoreleasing *)error
-{
-    if (!(self = [super initWithJSONDictionary:json error:error]))
-    {
-        return nil;
-    }
-    
-    [self initIdToken:error];
-    return self;
-}
-
-- (BOOL)initIdToken:(NSError *__autoreleasing *)error
-{
-    if (![NSString msidIsStringNilOrBlank:self.idToken])
-    {
-        self.idTokenObj = [[MSIDIdTokenClaims alloc] initWithRawIdToken:self.idToken error:error];
-        return self.idTokenObj != nil;
-    }
-    return YES;
-}
-
-- (NSInteger)expiresIn
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                          refreshToken:(MSIDBaseToken<MSIDRefreshableToken> *)token
+                                 error:(NSError **)error
 {
-    id expiresInObj = _json[MSID_OAUTH2_EXPIRES_IN];
-    NSInteger expiresIn = [MSIDHelpers msidIntegerValue:expiresInObj];
-    
-    if (!expiresIn && expiresInObj)
+    self = [self initWithJSONDictionary:json error:error];
+    if (self)
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelWarning,nil, @"Unparsable time - The response value for the access token expiration cannot be parsed: %@", expiresInObj);
+        if (token && [NSString msidIsStringNilOrBlank:_refreshToken])
+        {
+            _refreshToken = token.refreshToken;
+        }
     }
-    
-    return expiresIn;
-}
 
-- (void)setExpiresIn:(NSInteger)expiresIn
-{
-    NSString *expiresInString = [NSString stringWithFormat:@"%ld", (long)expiresIn];
-    _json[MSID_OAUTH2_EXPIRES_IN] = expiresInString;
+    return self;
 }
 
 - (NSDate *)expiryDate
 {
     NSInteger expiresIn = self.expiresIn;
-    
-    if (!expiresIn)
-    {
-        return nil;
-    }
-    
+
+    if (!expiresIn) return nil;
+
     return [NSDate dateWithTimeIntervalSinceNow:expiresIn];
 }
```

Comment:
Previously `MSIDTokenResponse` is supposed to be consumed immediately rather than being serialized/transported from one app to another app. 
If we want to serialize it and transport it from one app to another,  will it be better if we could also serialize the `derivedExpiredOn` value rather than just the `expiredIn` value? such that we don't have to worry about all possible delays or anything :P

Reply:
Good point, added `expires_on` to MSIDTokenResponse. `expires_on` has higher priority then `expires_in` now.